File: lib/rgfa/field/identifier_list_gfa2.rb
module RGFA::Field::IdentifierListGFA2

  def unsafe_decode(string)
    string.split(" ").map(&:to_sym)
  end

  def decode(string)
    validate_encoded(string)
    unsafe_decode(string)
  end

  def validate_encoded(string)
    if string !~ /^[ !-~]+$/
      raise RGFA::FormatError,
        "#{string.inspect} is not a valid list of GFA2 identifier\n"+
        "(it contains non-printable characters)"
    end
  end

  def validate_decoded(object)
    case object
    when Array
      object.each do |elem|
        elem = case elem
               when String
                 elem
               when Symbol
                 elem.to_s
               when RGFA::Line
                 elem.id.to_s
               else
                 raise RGFA::TypeError,
                   "the array contains an object of class #{elem.class}\n"+
                   "(accepted classes: String, Symbol, RGFA::Line)"
               end
        if elem !~ /^[!-~]+$/
          raise RGFA::FormatError,
          "the list contains an invalid GFA2 identifier (#{string.inspect})\n"+
          "(it contains spaces and/or non-printable characters)"
        end
      end
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: Array)"
    end
  end

  def validate(object)
    case object
    when String
      validate_encoded(object)
    when Array
      validate_decoded(object)
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: Array, String)"
    end
  end

  def unsafe_encode(object)
    case object
    when Array
      object.map do |elem|
        case elem
        when String, Symbol
          elem.to_s
        when RGFA::Line
          elem.id.to_s
        else
          raise RGFA::TypeError,
            "the array contains an object of class #{elem.class}\n"+
            "(accepted classes: String, Symbol, RGFA::Line)"
        end
      end.join(" ")
    when String
      return object
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: Array, String)"
    end
  end

  def encode(object)
    validate_decoded(object)
    return unsafe_encode(object)
  end

  module_function :decode
  module_function :unsafe_decode
  module_function :validate_encoded
  module_function :validate_decoded
  module_function :validate
  module_function :unsafe_encode
  module_function :encode

end

File: gfapy/field/identifier_list_gfa2.py
import gfapy
import re

def unsafe_decode(string):
  return string.split(" ")

def decode(string):
  validate_encoded(string)
  return unsafe_decode(string)

def validate_encoded(string):
  if not re.match("^[ !-~]+$", string):
    raise gfapy.FormatError(
      "{} is not a valid list of GFA2 identifier\n".format(repr(string))+
      "(it contains non-printable characters)")

def validate_decoded(obj):
  if isinstance(obj, list):
    for elem in obj:
      if isinstance(elem, gfapy.Line):
        elem = str(elem.name)
      elif not isinstance(elem, str):
        raise gfapy.TypeError(
          "the list contains an obj of class {}\n"
          .format(elem.__class__.__name__)+
          "(accepted classes: str, gfapy.Line)")
      if not re.match("^[!-~]+$", elem):
        raise gfapy.FormatError(
        "the list contains an invalid GFA2 identifier ({})\n"
        .format(repr(elem))+
        "(it contains spaces and/or non-printable characters)")
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__) +
      "(accepted classes: list)")

def unsafe_encode(obj):
  if isinstance(obj, list):
    def func(elem):
      if isinstance(elem, str):
        return elem
      elif isinstance(elem, gfapy.Line):
        return str(elem.name)
      else:
        raise gfapy.TypeError(
          "the list contains an obj of class {}\n"
          .format(elem.__class__.__name__)+
          "(accepted classes: str, gfapy.Line)")
    return " ".join(map(func, obj))
  elif isinstance(obj, str):
    return obj
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: list, str)")

def encode(obj):
  validate_decoded(obj)
  return unsafe_encode(obj)
